test/python/dialects/indexing/dialect.py

```diff
@@ -474,3 +477,153 @@ def testCanonicalizeTupleIndexCastListLiteral():
             str(t.type) == 'tensor<2x1xindex>' and t.is_constant() and
             np.array_equal(t.literal_value, [[0], [1]])
             for _, t in tens) and set(i for i, _ in tens) == set(tens_is)
+
+
+# CHECK-LABEL: TEST: testAdvancedIndexing
+@run
+def testAdvancedIndexing():
+  index = IndexType.get()
+  f32 = F32Type.get()
+  with mlir_mod_ctx() as module:
+    ten = Tensor.empty((10, 10, 10, 10), f32)
+    # CHECK: Tensor(%[[TEN:.*]], tensor<10x10x10x10xf32>)
+    print(ten)
+
+    w = ten[[[0], [1]], :, :, :]
+    idx_tensor_operand = Tensor(w.owner.operands[1])
+    # CHECK: %[[IDXTEN:.*]] tensor<2x1xindex> True
+    print(idx_tensor_operand.get_name(), idx_tensor_operand.type,
+          idx_tensor_operand.is_constant())
+    # CHECK{LITERAL}: [[0], [1]]
+    print(get_array_on_one_line(idx_tensor_operand.literal_value))
+    # CHECK: %{{.*}} = indexing.gather %[[TEN]][%[[IDXTEN]]] gather_dims([0]) unique : (tensor<10x10x10x10xf32>, tensor<2x1xindex>) -> tensor<2x10x10x10xf32>
+    print(w.owner)
+
+    w = ten[[[0], [1]], [[0], [1]], :, :]
```

Comment:
Would be good to lift some of these examples to the Tensor class doc comment and be pedantic about the flavor of IR being generated.

In particular here, the non-trivial part until you see it, I find, is how the `[[0], [1]], [[0], [1]]` indices get zipped/flattened to form the indexing tensor `[[0 0], [1 1]]`.

It seems one could question the syntax / behavior until one understands that this also allows you to specify non-contiguous gather dims which require a convention to handle the non contiguous indexing "holes".

Besides the ask for more handholding in the comments, I think this looks great!

Reply:
added a long doc string to `Tensor.__getitem__`.